### ISSUE TITLE
TopologicalOrderIterator even if the graph has cycles.

### ISSUE THREAD

**Issue by RahulNagekar** [jgrapht/jgrapht#750]
There could be cases where we want to carry out a topological order traversal even if cycles are present.
For example, let's say we want all vertices to be sorted in topological order, just there could be multiple valid topological paths as the graph has cycles.

Graph :   v1 --> v2 --> v3  --> v4 --> v2
In the case expected traversal/sort is : v1, v2, v3 ,v4 although if there is cycle with {v2,v3,v4} vertices.

The aim is to visit all the vertices in the most accurate topological order.

If there is already a way to do this could you please give me some pointers on this?

**Comment by d-michail** [comment]
The "most accurate topological order" sounds very application specific. You need to define it exactly in order to try and find solutions. 

As general pointers, there are the [Feedback vertex set](https://en.wikipedia.org/wiki/Feedback_vertex_set) and the [Feedback arc set](https://en.wikipedia.org/wiki/Feedback_arc_set) problems.

Read also about [layered graph drawing](https://en.wikipedia.org/wiki/Layered_graph_drawing).

In general these are hard problems, so you probably need to use some approximation or heuristic algorithms. In any case, it is probably best to try to tackle your specific problem instead of the general version.

If you end up implementing something along these lines, we welcome contributions as we are missing support for the feedback vertex set and edge problems.

**Comment by RahulNagekar** [comment]
The TopologicalOrderIterator in version  0.9.1 was doing what I am to proposing here. 
So, that made me think as a topological order can have valid n number of valid outcomes for Directed acyclic graph, then why can a cyclic graph also be traversed in topological order (which version 0.9.1) was already doing.

I thought about DFS and getting a starting node from the current TopologicalOrderIterator but then it won't work for a totally disjoint graph in the current graph.

So, basically I wonder why we can't topologically traverse a Simple directed graph having cycles?
All sources saying a graph being a DAG and then topologically sorting ...I am trying to see why such restriction. And if can disprove the restriction then we can implement the same :-)

**Comment by RahulNagekar** [comment]
For the example : Graph : v1 --> v2 --> v3 --> v4 --> v2
There is only 1 correct topological traversal i.e. v1 -> v2 -> v3 -> v4; as v1 has the lowest degree.

**Comment by d-michail** [comment]
A topological ordering exists if and only if a graph is a DAG. Our latest versions detect the cycle and throw an exception, while older versions did not.

In any case the documentation of older versions was correct clearly stating the following: 
```
For this iterator to work correctly the graph must be acyclic, and must not be modified
during iteration. Currently there are no means to ensure that, nor to fail-fast; the results
with cyclic input (including self-loops) or concurrent modifications are undefined. To
precheck a graph for cycles, consider using org.jgrapht.alg.CycleDetector or
org.jgrapht.alg.StrongConnectivityInspector.
```

The important term is "undefined" here which means that anything can happen, thus you cannot 
rely on the result. If your application is fine with that, consider using the old code.

**Comment by RahulNagekar** [comment]
Application wise I don't have any issue, also no issues with current implementation.
Just algorithm wise I am thinking topological traversing even a graph with cycles could become a concept and might also be useful in some practical problems.
As a cycle would actually violate the topological order definition, but again we can have an algorithm with the minimum number of violations and trying to do a topological sort.

I am not sure if there is a name to such an algorithm, I got this idea while comparing the older and newer version of the topological iterator.

I thought the idea is good so came out here to discuss it,

**Comment by RahulNagekar** [comment]
> The "most accurate topological order" sounds very application specific. You need to define it exactly in order to try and find solutions.
> 
> As general pointers, there are the [Feedback vertex set](https://en.wikipedia.org/wiki/Feedback_vertex_set) and the [Feedback arc set](https://en.wikipedia.org/wiki/Feedback_arc_set) problems.
> 
> Read also about [layered graph drawing](https://en.wikipedia.org/wiki/Layered_graph_drawing).
> 
> In general these are hard problems, so you probably need to use some approximation or heuristic algorithms. In any case, it is probably best to try to tackle your specific problem instead of the general version.
> 
> If you end up implementing something along these lines, we welcome contributions as we are missing support for the feedback vertex set and edge problems.

I got this reference https://stackoverflow.com/questions/17148103/topological-sort-of-cyclic-graph-with-minimum-number-of-violated-edges
And now I understand what you were suggesting, will try to understand this algo and will look towards contributing for the same.

Thanks :-)

**Comment by d-michail** [comment]
Actually the heuristic algorithm that is described in the stackoverflow answer: 

```
Eades, Peter, Xuemin Lin, and William F. Smyth. "A fast and effective heuristic for the
feedback arc set problem." Information Processing Letters 47.6 (1993): 319-323.
```

would be the perfect way to start as it is rather simple to implement.

**Comment by RahulNagekar** [comment]
Will try to understand it then, and as understood will implement. Thanks :-)
